### PR TITLE
feat: after dozen rewrrites - major ready

### LINKED ISSUE #1: todos
- isSync arg for handling sync/async in createErrorHelpers
- test for valibot tuple (picklist), if not tested
- fix `this['~zagora"]` in handler/handlerSync
- fix remaining `new Error` occurrences to `new ZagoraError`
- add `.options` for defining options schemas, append after handler args and before error helpers
- make input, output, and options schemas optional, default to z.any()-like schema or anySchema option in zagora config
- better handling if handler return tuple with length 2, maybe make error helpers return an instance so that it would be safer
  - anything returned is considered Result, if instance of Err or something like that is returned it us considered error

### PR DESCRIPTION
**Pull request by tunnckoCore**
close #1 

see more at #2 which was the base

---

## Pull request overview

This PR represents a major rewrite of the Zagora library (v1.0.0-next.5), transitioning from a class-based error handling approach to a more functional, result-oriented design. The library now provides type-safe, never-throwing procedures with comprehensive validation, context management, caching, and environment variable support.

**Key Changes:**
- Complete API redesign with builder pattern and fluent interfaces
- Migration from custom `ZagoraError` class to plain error objects with discriminated unions
- New result type: `{ ok: true, data } | { ok: false, error }` 
- Added `autoCallable` and `disableOptions` configuration options
- Environment variable validation with `.env()` method
- Built-in caching/memoization support
- Migration from Bun test to Vitest
- Comprehensive type-level testing with `expect-type`

## Summary by CodeRabbit

* **New Features**
  * Added public exports for error utilities, types, and helper functions
  * Restructured error handling framework with type guards and factories
  * Added support for environment variable validation and caching
  * Implemented autoCallable mode for streamlined handler invocation

* **Documentation**
  * Comprehensive README redesign with expanded API reference, examples, and feature showcase
  * Added simplified design documentation and agent guidelines

* **Chores**
  * Upgraded to version 1.0.0-next.5 with updated build tooling and CI/CD pipelines
  * Migrated testing framework and updated project configuration

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: test/main.test.ts [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+  }
+
+  fixture();
+  fixture(true);
+  fixture(true, true);
+  fixture(false, true);
+  fixture(false, true, true);
+  fixture(true, false);
+  fixture(true, true, false);
+  fixture(false, false, true);
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply test/main.test.ts]: wasn't really needed in my case, but solved.

File: src/utils.ts [hunk 2]
```diff
@@ -1,50 +1,396 @@
...
+      // ret.data is the result of the `cache.has`
+      return ret.data
+        ? tryCatch(() => cacheAdapter.get?.(key), false, "get")
+        : executeHandler(args, { handlerFn, cacheAdapter, key });
+    }
+    return ret;
+  }
+
+  return executeHandler(args, { handlerFn, cacheAdapter, key });
+}
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply src/utils.ts:115]: not too possible. `incoming` will never be circular, it's zod schemas.

but suggest me a serialization method - some JSON.stringify-like serializer without support for circular.

[the remark this answers is not in the document]
Comment (tunnckoCore) [reply src/utils.ts:115]: well, isn't ini the option 2, isn't that `JSON.stringify(args)` having the same problem if the passed procedure arguments are actually circular?

also, i don't think that's possible either, because they won't be valid (at runtime and comple-time) based on the schema.. so.. neither is a problem to begin with.

File: README.md [hunk 3]
```diff
@@ -1,214 +1,836 @@
...
+### Motivation
+While `orpc` is great and you can use it for direct function calls (and not network requests with `createRouterClient`), and for example for building "type-safe SDK"s, it does have a few opinions that may get in the way. I use it extensively in my projects, but `zagora` is smaller and even more focused approach - i always wanted "just functions" where you define input, outputs, and you get error-safe, typed function back not a wrapper around it.
+
+Both `tRPC` and `oRPC` are promoted as "backend", or specifically for when you're building "apps". Recently, all major frameworks also introduced similar concepts, like "server actions" and so on. All that is cool, but `zagora` is focused on building just functions, a low-level library for building other libraries - I have a lot of them, so i need a simple way for building type-safe and error-safe functions, where i don't necessarily need network layer and i don't need "routers" concept, and etc.
+
+They are built around the network, Zagora is built around functions with excellent egonomics, developer experience, and no assumptions. **It produces just functions, i cannot stress that enough.**
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply README.md]: "just functions" is absolute key. It's like "just works". It's not exactly an "only" thing, and it's not "exposes only functions" either.

fix the rest.

File: test/edge-cases.test.ts [hunk 4]
```diff
@@ -69,101 +142,340 @@ test("Zod tuple with multiple defaults", async () => {
...
+    }),
+  };
+
+  const func = zagora()
+    // no .input() call
+    .errors(errorSchemas)
+    .handler(({ errors }) => {
+      throw errors.BARRY_ERR({ msg: "Some custom error" });
+      // biome-ignore lint/correctness/noUnreachable: bruh
+      return "foo";
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply test/edge-cases.test.ts:322]: that is intentional, and commented there.

File: README.md [hunk 5]
```diff
@@ -1,214 +1,836 @@
...
+  .handler((_, num) => num * 2)
+  .callable();
+
+// TypeScript will warn you if you do not `await` here,
+// because the `has` method of the cache adapter is async,
+// thus the whole procedure becomes asynchronous.
+const res = await proc(22);
+```
+
+You can also provide the cache through `.callable({ cache })`. That is useful, if you want to provide it at "execution place", not at "definition place". For example, you'd have a set of procedures written at one place, then through "router" or some object that combiens them you want to call them at a `Request/Response` server handler.
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply README.md:647]: okay, just fucking stop suggesting me typos and tweaks every fucking time, you could do it all at fucking once.

gtfo

File: test/edge-cases.test.ts [hunk 6]
```diff
@@ -1,7 +1,46 @@
...
+    .errors(errorSchemas)
+    // .errors({ FOO: z.object({ msg: z.string() }) })
+    .handler(({ errors }) => {
+      throw errors.BAR_ERR({ msg: "Custom error" });
+
+      // NOTE: simulate return, because the return types break
+      // when no return and no `input` schema
+
+      // biome-ignore lint/correctness/noUnreachable: bruh
+      return "foo";
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply test/edge-cases.test.ts:26]: that is fucking intentional, and already commented there.

gtfo
